solidity/random-beacon/contracts/ReimbursmentPool.sol

```diff
@@ -0,0 +1,150 @@
+//
+// ▓▓▌ ▓▓ ▐▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▄
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▌▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓    ▓▓▓▓▓▓▓▀    ▐▓▓▓▓▓▓    ▐▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓         ▓▓▓▓▓▓▓▓▓▓         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+//   ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+//   ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+// ▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+// ▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+//
+//                           Trust math, not hardware.
+
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+contract ReimbursementPool is Ownable, ReentrancyGuard {
+    /// @notice Authorized contracts that can interact with the reimbursment pool.
+    ///         Authorization can be granted and removed by the owner.
+    mapping(address => bool) public isAuthorized;
+
+    /// @notice Static gas includes:
+    ///         - cost of the refund function
+    ///         - base transaction cost
+    uint256 public staticGas;
+
+    /// @notice Max gas price used to reimburse a transaction submitter. Protects
+    ///         against malicious operator-miners.
+    uint256 public maxGasPrice;
+
+    event StaticGasUpdated(uint256 newStaticGas);
+
+    event MaxGasPriceUpdated(uint256 newMaxGasPrice);
+
+    event SendingEtherFailed(uint256 refundAmount, address receiver);
+
+    event AuthorizeContract(address thirdPartyContract);
+
+    event UnauthorizeContract(address thirdPartyContract);
+
+    event FundsWithdrawn(uint256 withdrawnAmount, address receiver);
+
+    constructor(uint256 _staticGas, uint256 _maxGasPrice) {
+        staticGas = _staticGas;
+        maxGasPrice = _maxGasPrice;
+    }
+
+    /// @notice Receive ETH
+    receive() external payable {}
+
+    /// @notice Refunds ETH to a spender for executing specific transactions.
+    /// @dev Ignoring the result of sending ETH to a receiver is made on purpose.
+    ///      For EOA receiving ETH should always work. If a receiver is a smart
+    ///      contract, then we do not want to fail a transaction, because in some
+    ///      cases the refund is done at the very end of multiple calls where all
+    ///      the previous calls were already paid off. It is a receiver's smart
+    ///      contract resposibility to make sure it can receive ETH.
+    /// @dev Only authorized contracts are allowed calling this function.
+    /// @param gasSpent Gas spent on a transaction that needs to be reimbursed.
+    /// @param receiver Address where the reimbursment is sent.
+    function refund(uint256 gasSpent, address receiver) external nonReentrant {
```

Comment:
What is the gas cost difference with/without non-reentrant? All approved contracts are supposed to be trusted.

Reply:
With `nonReentrant`: 40433  ·      41893  ·      41075 
Without:  38083  ·      39543  ·      38725

If per prev. comments we take `max`, then the diff is `3168`. 

Yes, there are supposed to be trusted, but what if this trusted contract has a bug that can enable reentrancy here? The diff is not that small, but also not too big compared to the overall cost and having in mind the risk of draining the pool maybe we should leave the `nonReentrant` modifier in place?